File: src/main/kotlin/io/titandata/titan/providers/kubernetes/Uninstall.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.kubernetes

import io.titandata.client.apis.RepositoriesApi
import io.titandata.titan.clients.Docker
import io.titandata.titan.providers.Kubernetes
import io.titandata.titan.utils.CommandExecutor
import io.titandata.titan.utils.ProgressTracker

class Uninstall (
        private val titanServerVersion: String,
        private val exit: (message: String, code: Int) -> Unit,
        private val remove: (container: String, force: Boolean) -> Unit,
        private val commandExecutor: CommandExecutor = CommandExecutor(),
        private val docker: Docker = Docker(commandExecutor, Kubernetes.Identity),
        private val repositoriesApi: RepositoriesApi = RepositoriesApi(),
        private val track: (title: String, function: () -> Any) -> Unit = ProgressTracker()::track
) {
    fun uninstall(force: Boolean) {
        if (docker.titanServerIsAvailable()) {
            val repositories = repositoriesApi.listRepositories()
            for (repo in repositories) {
                if (!force) {
                    exit("repository ${repo.name} exists, remove first or use '-f'", 1)
                }
                remove(repo.name, true)
            }
        }
        if (docker.titanServerIsAvailable()) docker.rm("${docker.identity}-server", true)
        track ("Removing titan-data Docker volume") {
            docker.removeVolume("titan-data")
        }
        track ("Removing Titan Docker image") {
            docker.removeTitanImages(titanServerVersion)
        }
        println("Uninstalled titan infrastructure")
    }
}

File: internal/app/providers/kubernetes/Uninstall.go
package kubernetes

import (
	"fmt"
	"github.com/briandowns/spinner"
	"os"
	"strconv"
	"time"
	"titan/internal/app/clients"
)

func Uninstall(force bool, removeImages bool, context string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)
	docker := clients.Docker(context, port)

	available, _ :=  docker.TitanServerIsAvailable()
	if available {
		repos, _, _ := repositoriesApi.ListRepositories(ctx)
		for _, repo := range repos {
			if !force {
				fmt.Println("repository" + repo.Name + "exists, remove first or use '-f'")
				os.Exit(1)
			}
		}
		docker.Remove("titan-" + context + "-server", true)
	}

	s := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
	s.HideCursor = true

	s.Prefix = "Removing Titan Docker volume "
	s.FinalMSG = "Titan Docker volume removed"
	s.Start()
	docker.RemoveVolume("titan-" + context + "-date", true)
	s.Stop()
	fmt.Println()

	if removeImages {
		s.Prefix = "Removing Titan Docker image "
		s.FinalMSG = "Titan Docker image removed"
		s.Start()
		docker.RemoveTitanImages("latest")
		s.Stop()
		fmt.Println()
	}
	fmt.Println("Uninstalled titan infrastructure")
}